File: backend/app/services/token_calculator.py
import math
from typing import Dict, Any

class TokenCalculator:
    """
    Calculates token counts for prompts and completions,
    and estimates token savings compared to standard conversational AI chatter.
    """
    
    @staticmethod
    def estimate_tokens(text: str) -> int:
        """
        Fast token estimator (4 characters ~= 1 token on average for standard English/code).
        """
        if not text:
            return 0
        words = len(text.split())
        chars = len(text)
        # Hybrid formula combining word count and char count for higher precision
        return max(1, math.ceil((words * 1.3 + chars / 4) / 2))

    @classmethod
    def calculate_metrics(
        self,
        wrapped_system_prompt: str,
        user_query: str,
        ai_response: str,
        raw_prompt_tokens: int = 0,
        raw_completion_tokens: int = 0
    ) -> Dict[str, int]:
        """
        Calculates prompt, completion, total, and estimated tokens saved.
        
        Standard Conversational AI baseline:
        - Includes verbose greetings, explanations, pleasantries, step-by-step preamble (~350-800 extra tokens).
        - Conversational responses usually average ~600-1200 tokens.
        - Hyper-efficient command blueprints compress outputs by ~60-75%.
        """
        prompt_tokens = raw_prompt_tokens if raw_prompt_tokens > 0 else self.estimate_tokens(wrapped_system_prompt + user_query)
        completion_tokens = raw_completion_tokens if raw_completion_tokens > 0 else self.estimate_tokens(ai_response)
        total_tokens = prompt_tokens + completion_tokens

        # Baseline calculation for un-wrapped conversational AI response to the same query
        # Typical conversational AI chatter adds ~250 tokens in intro/outro + ~1.8x verbose explanation text
        baseline_completion_tokens = math.ceil(completion_tokens * 2.2 + 250)
        baseline_prompt_tokens = math.ceil(prompt_tokens * 1.5 + 150) # Standard multi-turn verbose prompt context
        baseline_total_tokens = baseline_prompt_tokens + baseline_completion_tokens

        estimated_tokens_saved = max(0, baseline_total_tokens - total_tokens)

        return {
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "total_tokens": total_tokens,
            "baseline_total_tokens": baseline_total_tokens,
            "estimated_tokens_saved": estimated_tokens_saved
        }

File: frontend/src/services/tokenCalculator.js
// Senior Developer Token & Financial Cost Savings Engine

// Estimated API Cost per 1k tokens (USD)
const API_RATES = {
  openai: {
    input: 0.00015,  // $0.15 per 1M tokens (gpt-4o-mini)
    output: 0.00060  // $0.60 per 1M tokens
  },
  anthropic: {
    input: 0.00025,  // $0.25 per 1M tokens (claude-3-5-haiku)
    output: 0.00125  // $1.25 per 1M tokens
  },
  gpt4: {
    input: 0.0025,   // $2.50 per 1M tokens (gpt-4o baseline conversational)
    output: 0.0100   // $10.00 per 1M tokens
  }
};

export function estimateTokenCount(text) {
  if (!text) return 0;
  const words = text.split(/\s+/).filter(Boolean).length;
  const chars = text.length;
  return Math.max(1, Math.ceil((words * 1.35 + chars / 4) / 2));
}

export function calculateFinancialSavings(tokensSaved, provider = 'openai') {
  const rate = API_RATES[provider] || API_RATES.openai;
  // Blend input & output savings estimation
  const blendedRatePerToken = (rate.input * 0.4 + rate.output * 0.6);
  const dollarsSaved = (tokensSaved * blendedRatePerToken);
  
  return {
    dollarsSaved: Math.max(0, Math.round(dollarsSaved * 10000) / 10000),
    formattedDollars: dollarsSaved < 0.01 ? `< $0.01` : `$${dollarsSaved.toFixed(3)}`
  };
}

export function computeMetrics(systemPrompt, userQuery, aiResponse, rawPromptTokens = 0, rawCompletionTokens = 0, provider = 'openai') {
  const promptTokens = rawPromptTokens > 0 ? rawPromptTokens : estimateTokenCount(systemPrompt + userQuery);
  const completionTokens = rawCompletionTokens > 0 ? rawCompletionTokens : estimateTokenCount(aiResponse);
  const totalTokens = promptTokens + completionTokens;

  // Conversational AI chatter baseline (verbose greetings, multi-turn intro/outro preamble)
  const baselineCompletion = Math.ceil(completionTokens * 2.4 + 300);
  const baselinePrompt = Math.ceil(promptTokens * 1.6 + 180);
  const baselineTotal = baselinePrompt + baselineCompletion;

  const tokensSaved = Math.max(120, baselineTotal - totalTokens);
  const reductionPercentage = Math.round((tokensSaved / baselineTotal) * 1000) / 10;
  const financial = calculateFinancialSavings(tokensSaved, provider);

  return {
    prompt_tokens: promptTokens,
    completion_tokens: completionTokens,
    total_tokens: totalTokens,
    baseline_tokens: baselineTotal,
    tokens_saved: tokensSaved,
    efficiency_percentage: Math.min(92.0, Math.max(50.0, reductionPercentage)),
    dollars_saved: financial.dollarsSaved,
    formatted_dollars: financial.formattedDollars
  };
}
